web/src/types.ts

```diff
@@ -6,19 +6,111 @@ export type MatchedMessage = {
 export type MessageType = "private" | "stream";
+export type MessageContentType = "text/html" | "text/x-markdown";
+export type MessageReactionType = "unicode_emoji" | "realm_emoji" | "zulip_extra_emoji";
+
+// TODO/typescript: Move this to server_events
+export type TopicLink = {
+    text: string;
+    url: string;
+};
+
+// TODO/typescript: Move these types to message_store
+
+export type DisplayRecipientUser = {
+    email: string;
+    full_name: string;
+    id: number;
+    is_mirror_dummy: boolean;
+    unknown_local_echo_user?: boolean;
+};
+
+export type DisplayRecipient = string | DisplayRecipientUser[];
+
+export type MessageEditHistoryEntry = {
+    user_id: number | null;
+    timestamp: number;
+    prev_content?: string;
+    prev_rendered_content?: string;
+    prev_rendered_content_version?: number;
+    prev_stream?: number;
+    prev_topic?: string;
+    stream?: number;
+    topic?: string;
+};
+
+export type MessageReaction = {
+    emoji_name: string;
+    emoji_code: string;
+    reaction_type: MessageReactionType;
+    user_id: number;
+    user: DisplayRecipientUser;
+};
 
 // TODO/typescript: Move this to message_store
 export type RawMessage = {
+    avatar_url: string;
+    client: string;
+    content: MessageContentType;
+    content_type: string;
+    display_recipient: DisplayRecipient;
+    edit_history?: MessageEditHistoryEntry[];
+    id: number;
+    is_me_message: boolean;
+    last_edit_timestamp?: number;
+    reactions: MessageReaction[];
+    recipient_id: number;
     sender_email: string;
-    stream_id: number;
+    sender_full_name: string;
+    sender_id: number;
+    sender_realm_str: string;
+    stream_id?: number;
     subject: string;
+    submessages: string[];
+    timestamp: number;
+    topic_links: TopicLink[];
     type: MessageType;
+    flags: string[];
 } & MatchedMessage;
 
 // TODO/typescript: Move this to message_store
-export type Message = RawMessage & {
-    to_user_ids: string;
+export type MessageCleanReaction = {
+    class: string;
+    count: number;
+    emoji_alt_code: string;
+    emoji_code: string;
+    emoji_name: string;
+    is_realm_emoji: boolean;
+    label: string;
+    local_id: string;
+    reaction_type: string;
+    user_ids: number[];
+    vote_text: string;
+};
+
+// TODO/typescript: Move this to message_store
+export type Message = Omit<RawMessage, "reactions"> & {
+    alerted: boolean;
+    clean_reactions: Map<string, MessageCleanReaction>;
+    collapsed: boolean;
+    display_reply_to?: string;
+    historical: boolean;
+    is_private?: boolean;
+    is_stream?: boolean;
+    mentioned: boolean;
+    mentioned_me_directly: boolean;
+    message_reactions: MessageCleanReaction[];
+    pm_with_url?: string;
+    reply_to: string;
+    sent_by_me: boolean;
+    starred: boolean;
+    starred_status: string;
+    stream?: string;
+    unread: boolean;
+    url: string;
     topic: string;
+    to_user_ids: string;
```

Comment:
`to_user_ids` only appears in direct messages

Reply:
Made it optional 👍